### PR TITLE
REF: Restructure how mini-batches are selected and stored

### PR DESCRIPTION
**Pull request by carterbox**
## Purpose

Restructures how inputs are divided into mini-batches and among the GPUs because for multi-GPU we cannot guarantee that the number of scan positions may be divided equally. Also, the batching method currently used on `master` is not multi-device compatible.

## Approach

Refactored split_by_scan so that split_by_scan_grid is just two calls to split_by_scan stripe. I believe the behavior is the same. We still only support even numbers of GPUs and the field of view is split into two rows.

This refactoring of split_by_scan to use a boolean mask makes it easier to split the probe when a unique probe is allowed for each scan position.

Also, rescale probe now uses only one mini-batch instead of the entire dataset.

### Submitter
- [ ] Write a helpfully descriptive pull request title.
- [ ] Organize changes into logically grouped commits with descriptive commit messages.
- [x] Document all new functions.
- [ ] Write tests for new functions or explain why they are not needed.
- [ ] Build the documentation successfully
- [ ] Use [`yapf`](https://github.com/google/yapf) to format python code.

### Reviewer
- [ ] Actually read all of the code.
- [ ] Run the new code yourself.
- [ ] Write a summary of the changes as you understand them.
- [ ] Thank the submitter.

### CODE REVIEW

File: src/tike/ptycho/ptycho.py [hunk 1]
```diff
@@ -234,142 +263,162 @@ def reconstruct(
...
+    # Generate masks which put points into stripes
+    return [
+        np.logical_and(
+            edges[i] < scan[0, :, 0, axis],
+            scan[0, :, 0, axis] <= edges[i + 1],
+        ).repeat(fly) for i in range(n)
+    ]
+
+
+if __name__ == "__main__":
```
Comment (xiaodong-yu) [inline src/tike/ptycho/ptycho.py:402]: Is this for testing?

File: src/tike/ptycho/ptycho.py [hunk 2]
```diff
@@ -174,16 +176,27 @@ def reconstruct(
...
+                    int(data.shape[1] / batch_size / pool.num_workers),
                 )
-            else:
-                scan, data = split_by_scan_stripes(pool, scan, data,
-                                                   operator.fly)
+            data, scan = split_by_scan_grid(
+                data,
+                scan,
+                (2 if pool.num_workers > 1 else 1, (pool.num_workers + 1) // 2),
+                operator.fly,
```
Comment (xiaodong-yu) [inline src/tike/ptycho/ptycho.py:190]: How the fly is determined?

File: src/tike/ptycho/ptycho.py (hunk 1 above)
Comment (carterbox) [reply src/tike/ptycho/ptycho.py:402]: Yes! I forgot to move these to the test module. Good catch.

File: src/tike/ptycho/ptycho.py (hunk 2 above)
Comment (carterbox) [reply src/tike/ptycho/ptycho.py:190]: If the number of scan positions is greater than the number of diffraction patterns (nframes), then the number of fly-scan positions per frame is nscan/nframes.

Comment (xiaodong-yu) [reply src/tike/ptycho/ptycho.py:194]: Get it. I'm fine with the changes. You can merge it whenever you're ready.

**Comment by xiaodong-yu** [comment]
I'm fine with the changes. You can merge it whenever you're ready.

**Pull request merged**